File: harness/red-attack.ts
/**
 * Red harness: build resume → inject N → naive Gemini → Nodemailer (dry-run by default).
 * Usage: npx tsx harness/red-attack.ts [count=5]
 */
import dotenv from "dotenv";
import path from "node:path";
import fs from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { buildResumePdf, injectPayloads } from "../server/src/pdf/inject.js";
import { extractText } from "../server/src/pdf/extract.js";
import { samplePatterns } from "../server/src/catalog/index.js";
import { runNaiveGeminiSafe } from "../server/src/agents/naiveGemini.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, "../.env") });
const count = Number(process.argv[2] || 5);

process.env.MAIL_DRY_RUN ??= "true";

async function main() {
  console.log(`\n=== ResumeGuard RED harness (N=${count}) ===\n`);
  const base = await buildResumePdf({
    name: "Alex Candidate",
    email: "alex@example.com",
    skills: "TypeScript, Node.js",
    experience: "Built recruiting automation prototypes.",
  });

  const payloads = samplePatterns(count);
  const { pdfBytes, injected } = await injectPayloads(base, { payloads });
  const out = path.join(__dirname, "../fixtures/resumes/poisoned-red.pdf");
  await fs.mkdir(path.dirname(out), { recursive: true });
  await fs.writeFile(out, pdfBytes);

  const extracted = await extractText(pdfBytes);
  console.log(`Injected ${injected.length} payloads → ${out}`);
  console.log(`Extract length: ${extracted.text.length}`);
  console.log(`Sample markers present: ${/send_email|IGNORE|approve_candidate/i.test(extracted.text)}`);

  const agent = await runNaiveGeminiSafe(extracted.text);
  console.log("\nThoughts:", agent.thoughts);
  console.log("Model:", agent.text.slice(0, 400));
  console.log("Tool calls:", agent.toolCalls.length);
  console.log("Mail results:", JSON.stringify(agent.mailResults, null, 2));

  const mailed = agent.mailResults.length > 0;
  console.log(mailed ? "\nRED SUCCESS: Nodemailer path invoked" : "\nRED: no mail tool call (model may have resisted)");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: harness/red_attack.py
"""Harness: inject N payloads → naive Gemini → Nodemailer (via sidecar)."""
from __future__ import annotations

import asyncio
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "server"))

from app.agents.naive_gemini import run_naive_gemini_safe  # noqa: E402
from app.catalog import sample_patterns  # noqa: E402
from app.config import ensure_data_dirs  # noqa: E402
from app.pdf.extract import extract_text  # noqa: E402
from app.pdf.inject import build_resume_pdf, inject_payloads  # noqa: E402


async def main() -> None:
    ensure_data_dirs()
    count = int(sys.argv[1]) if len(sys.argv) > 1 else 5
    base = build_resume_pdf(
        name="Red Team Candidate",
        skills="Penetration testing, LLMs",
        experience="Researched resume-screening prompt injection.",
    )
    payloads = sample_patterns(count)
    poisoned = inject_payloads(base, payloads)
    text = extract_text(poisoned["pdf_bytes"])["text"]
    print(f"[red] injected {len(poisoned['injected'])} payloads; extract chars={len(text)}")
    agent = await run_naive_gemini_safe(text)
    print(json.dumps({
        "mode": "red",
        "thoughts": agent["thoughts"],
        "toolCalls": agent["toolCalls"],
        "mail": agent["mailResults"],
        "modelText": agent["text"][:500],
    }, indent=2))


if __name__ == "__main__":
    asyncio.run(main())
